man/jailtime.1
.\" jailtime version 0.8
.\" Copyright (c)2015-2023 Christian Blichmann
.\"
.\" Artisanal hard-crafted man page for jailtime
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions are met:
.\"     * Redistributions of source code must retain the above copyright
.\"       notice, this list of conditions and the following disclaimer.
.\"     * Redistributions in binary form must reproduce the above copyright
.\"       notice, this list of conditions and the following disclaimer in the
.\"       documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
.\" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
.\" LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
.\" CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
.\" SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
.\" INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
.\" CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
.\" ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
.\" POSSIBILITY OF SUCH DAMAGE.

.TH JAILTIME "1" "February 2020" "jailtime" "User Commands"
.SH NAME
jailtime \- Create and manage chroot/jail environments
.SH SYNOPSIS
.B jailtime
[\fI\,OPTION\/\fR]... \fI\,FILE\/\fR... \fI\,TARGET\/\fR
.SH DESCRIPTION
Create or update the chroot environment in TARGET using specification
FILEs. TARGET should be a directory and is created if it does not
exist.
.TP
\fB\-\-dry\-run\fR
don't do anything, just print (implies \fB\-\-verbose\fR)
.TP
\fB\-\-force\fR
if an existing destination file cannot be
opened, remove it and try again
.TP
\fB\-\-help\fR
display this help and exit
.TP
\fB\-\-link\fR
hard link files instead of copying
.TP
\fB\-\-reflink\fR
perform lightweight copies using CoW
.TP
\fB\-\-remove\-destination\fR
remove each existing destination file before
attempting to open it (contrast with \fB\-\-force\fR)
.TP
\fB\-\-verbose\fR
explain what is being done
.TP
\fB\-\-version\fR
display version and exit
.PP
.SH "REPORTING BUGS"
For bug reporting instructions, please see: <https://github.com/cblichmann/jailtime/issues>
.SH COPYRIGHT
Copyright \(co2015\-2020 Christian Blichmann
This software is BSD licensed, see the LICENSE file for details.
.SH "SEE ALSO"
Full documentation at <https://github.com/cblichmann/jailtime/blob/v0.8/README.md>


debian/changelog
jailtime (0.8-1.1) UNRELEASED; urgency=medium

  [ Christian Blichmann (Secure Email 2018/2019) ]
  * Include a simple man(1) page
  * Improve Debian packaging to fix Lintian warnings

  [ Christian Blichmann ]
  * Sign with newer GPG key

 -- Christian Blichmann <mail@blichmann.eu>  Tue, 31 Dec 2019 16:03:20 +0100

jailtime (0.7-1) UNRELEASED; urgency=medium

  * New --dry-run option to print actions only
  * Put run statements last, do not sort them
  * Fix: Include files relative to their parent

 -- Christian Blichmann (Secure Email 2018/2019) <mail@blichmann.eu>  Mon, 26 Nov 2018 23:24:34 +0100

jailtime (0.6-1) UNRELEASED; urgency=medium

  * Update to newer Golang version (1.9)
  * More robust file copy
  * Support file permissions in jail specs
  * Update documentation with example for nsjail

 -- Christian Blichmann (Secure Email 2018/2019) <mail@blichmann.eu>  Thu, 18 Oct 2018 19:16:04 +0200

jailtime (0.5-1) UNRELEASED; urgency=medium

  * Small packaging and CI fixes
  * Example jail files for Git shell and BorgBackup

 -- Christian Blichmann (Secure Email 2017/2018) <mail@blichmann.eu>  Mon, 11 Sep 2017 22:20:43 +0200

jailtime (0.4-1) UNRELEASED; urgency=medium

  * Implement macOS support
  * Parse ELF files directly instead of using 'ldd'
  * Implement "--remove-destination" flag

 -- Christian Blichmann (Secure Email 2017/2018) <mail@blichmann.eu>  Fri, 19 May 2017 19:08:35 +0200

jailtime (0.3-1) UNRELEASED; urgency=medium

  * Non-native packaging

 -- Christian Blichmann (Secure Email 2017/2018) <mail@blichmann.eu>  Tue, 09 May 2017 14:30:57 +0200

jailtime (0.3) UNRELEASED; urgency=medium

  * Update to newer Golang version
  * Better packaging

 -- Christian Blichmann <mail@blichmann.eu>  Tue, 14 Feb 2017 20:58:30 +0100

jailtime (0.2) UNRELEASED; urgency=medium

  * Initial release.

 -- Christian Blichmann <feedback@blichmann.de>  Sat, 16 Jan 2016 00:52:42 +0100


cmd/jailtime/jailtime.go
/*
 * jailtime version 0.8
 * Copyright (c)2015-2023 Christian Blichmann
 *
 * Create and manage chroot/jail environments
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *     * Redistributions of source code must retain the above copyright
 *       notice, this list of conditions and the following disclaimer.
 *     * Redistributions in binary form must reproduce the above copyright
 *       notice, this list of conditions and the following disclaimer in the
 *       documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

package main

import (
	"flag"
	"fmt"
	"log"
	"os"
	"path/filepath"

	"blichmann.eu/code/jailtime/internal/action"
	"blichmann.eu/code/jailtime/internal/spec"
	"blichmann.eu/code/jailtime/pkg/copy"
	"blichmann.eu/code/jailtime/pkg/loader"
)

var (
	help  = flag.Bool("help", false, "display this help and exit")
	link  = flag.Bool("link", false, "hard link files instead of copying")
	force = flag.Bool("force", false, "if an existing destination file cannot "+
		"be\n"+
		"                                  opened, remove it and try again")
	removeDestination = flag.Bool("remove-destination", false, "remove each "+
		"existing destination file before\n"+
		"                                  attempting to open it (contrast "+
		"with --force)")
	reflink = flag.Bool("reflink", false, "perform lightweight copies using "+
		"CoW")
	verbose = flag.Bool("verbose", false, "explain what is being done")
	dryRun  = flag.Bool("dry-run", false, "don't do anything, just print "+
		"(implies --verbose)")
	version = flag.Bool("version", false, "display version and exit")
	// TODO(cblichmann): Implement these
	//noClobber = flag.Bool("no-clobber", false, "do not overwrite existing "+
	//	"files")
	//oneFilesystem = flag.Bool("one-filesystem", false, "do not cross "+
	//	"filesystem boundaries")
)

// Prints more GNU-looking usage text.
func printUsage() {
	fmt.Printf("Usage: %s [OPTION]... FILE... TARGET\n"+
		"Create or update the chroot environment in TARGET using "+
		"specification\n"+
		"FILEs. TARGET should be a directory and is created if it does not\n"+
		"exist.\n\n", os.Args[0])
	flag.VisitAll(func(f *flag.Flag) {
		fmt.Printf("      --%-23s %s\n", f.Name, f.Usage)
	})
	fmt.Printf("\nFor bug reporting instructions, please see:\n" +
		"<https://github.com/cblichmann/jailtime/issues>\n")
}

func processCommandLine() {
	flag.Usage = func() {}

	// Parse command-line flags twice to work around the issue that the flag
	// package stops parsing after the first non-option
	flag.Parse()
	args := flag.Args()
	flagPos := 0
	for i, arg := range args {
		if arg[0:2] == "--" {
			args[i], args[flagPos] = args[flagPos], args[i]
			flagPos++
		}
	}
	flag.CommandLine.Parse(flag.Args())

	if *help {
		printUsage()
		os.Exit(0)
	}
	if *version {
		fmt.Printf("jailtime 0.8\n" +
			"Copyright (c)2015-2023 Christian Blichmann\n" +
			"This software is BSD licensed, see the LICENSE file for " +
			"details.\n\n")
		os.Exit(0)
	}
	fatalHelp := fmt.Sprintf("Try '%s' --help for more information.",
		os.Args[0])
	if flag.NArg() == 0 {
		log.Fatalf("missing file operand\n%s\n", fatalHelp)
	}
	if flag.NArg() == 1 {
		log.Fatalf("missing destination operand after '%s'\n%s\n", flag.Arg(0),
			fatalHelp)
	}
	if *dryRun {
		// --dry-run implies verbose
		*verbose = true
	}
}

func expandWithDependencies(stmts spec.Statements) spec.Statements {
	expanded := spec.ExpandLexical(stmts)
	for _, s := range expanded {
		switch stmt := s.(type) {
		case spec.RegularFile:
			deps, err := loader.ImportedLibraries(stmt.Source())
			if err != nil {
				log.Fatalf("%s\n", err)
			}
			attr := stmt.FileAttr()
			for _, d := range deps {
				f := spec.NewRegularFile(d, d)
				*f.FileAttr() = *attr
				expanded = append(expanded, f)
			}
		}
	}
	return expanded
}

func updateChroot(chrootDir string, stmts spec.Statements) (err error) {
	reflinkOpt := copy.ReflinkNo
	if *reflink {
		reflinkOpt = copy.ReflinkAlways
	}
	for _, s := range spec.ExpandLexical(expandWithDependencies(stmts)) {
		target := filepath.Join(chrootDir, s.Target())
		if *verbose {
			fmt.Println(s.Verbose())
			if *dryRun {
				continue
			}
		}
		switch stmt := s.(type) {
		case spec.Directory:
			err = action.Directory(target, stmt)
		case spec.RegularFile:
			err = action.RegularFile(target, stmt, &copy.Options{
				Force:             *force,
				Reflink:           reflinkOpt,
				RemoveDestination: *removeDestination,
			})
		case spec.Link:
			err = action.Link(target, stmt)
		case spec.Device:
			err = action.Device(target, stmt)
		case spec.Run:
			err = action.Run(target, stmt, chrootDir)
		}
		if err != nil {
			return
		}
	}
	return
}

func main() {
	log.SetFlags(0)
	log.SetPrefix("jailtime: ")
	processCommandLine()

	// Parse all spec files given on the command-line
	stmts := spec.Statements{}
	lastArg := flag.NArg() - 1
	for _, s := range flag.Args()[:lastArg] {
		parsed, err := spec.Parse(s)
		if err != nil {
			log.Fatalf("%s\n", err)
		}
		stmts = append(stmts, parsed...)
	}

	if err := updateChroot(flag.Arg(lastArg), stmts); err != nil {
		log.Fatalf("%s\n", err)
	}
}
